Created an expression runner for Spark functions

codyschierbeck opened the pull request:
velox_expression_runner_test is only linked to velox_functions_prestosql. 
Added a spark_expression_runner_test that is linked to velox_functions_spark to allow for running SparkSQL functions.

See comments for examples.

codyschierbeck commented:
Examples:

![image](https://github.com/facebookincubator/velox/assets/14911410/7eb0efc2-dc71-454e-a068-4da9dedbddb0)

I have been implementing SparkSQLs decode function and some additional Spark JSON functions on the side. Merged all of these into a testing branch based on top of this branch to make examples. In the example you can see:

1. to_json, which matches PrestoSQLs parse_json.
2. get_json_object which matches PrestoSQLs json_extract_scalar
3. decode, matching the example function found here: https://spark.apache.org/docs/latest/api/sql/index.html#decode
4. regexp_replace

codyschierbeck commented:
@mbasmanova 

I discovered some weird behavior in make format-fix. It seems like it is ignoring velox/expression/tests/ExpressionRunnerTest.cpp. I assumed I didn't need to check formatting with SparkExpressionRunnerTest.cpp since it was almost a 1:1 copy of ExpressionRunnerTest.cpp. Attached image to explain:


![image](https://github.com/facebookincubator/velox/assets/14911410/74982120-a543-4cd0-9c5e-bee562d34992)

On the left is ExpressionRunnerTest.cpp, and on the right are the changes to SparkExpressionRunnerTest.cpp after running make format-fix. No changes were ExpressionRunnerTest.cpp were suggested.

kgpai commented on velox/docs/develop/testing/fuzzer.rst:
```diff
@@ -267,7 +267,7 @@ evaluate constant expression or together with ``--input_path`` to evaluate
 exclusive. If both are specified, ``--sql`` is used while ``--sql_path`` is
 ignored. ``--sql`` option allow to specify multiple comma-separated SQL
-expressions.
+expressions. To run SparkSQL queries, use velox/expression/tests:spark_expression_runner_test instead of velox_expression_runner_test.
```
Nit can we add this on another paragraph with a small title and some documentation on flags etc.

kgpai commented on velox/expression/tests/CMakeLists.txt:
```diff
@@ -115,11 +115,20 @@ add_library(velox_expression_runner ExpressionRunner.cpp)
 
+add_executable(spark_expression_runner_test SparkExpressionRunnerTest.cpp)
+target_link_libraries(
+  spark_expression_runner_test spark_expression_runner velox_exec_test_lib
```
looks like spark_expression_runner library is only used for spark_expression_runner_test, why not put them all as a part of spark_expression_runner_test then.

kgpai commented on velox/expression/tests/SparkExpressionRunnerTest.cpp:
```diff
@@ -0,0 +1,195 @@
+
+using namespace facebook::velox;
+
+DEFINE_string(
```
These are the same as the ones present in presto expression fuzzer if I am not mistaken, maybe we can move the two to a common place rather than have 2 instances of same thing ?

kgpai commented on velox/expression/tests/SparkExpressionRunnerTest.cpp:
```diff
@@ -0,0 +1,195 @@
+    "indices that specify which columns of the input row vector should "
+    "be wrapped in lazy.");
+
+static bool validateMode(const char* flagName, const std::string& value) {
```
Same here..

kgpai commented on velox/expression/tests/SparkExpressionRunnerTest.cpp:
```diff
@@ -0,0 +1,195 @@
+    VELOX_CHECK(!sql.empty());
+  }
+
+  functions::sparksql::registerFunctions("");
```
The only major difference with Presto expression fuzzer is the fact that the registered functions are different, will it make more sense to actually add a flag there to choose spark or presto registry rather than create this executable ?

codyschierbeck replied:
@kgpai 

I have these changes staged, please let me know what you think.

**Flag registration**

```cpp
DEFINE_string(
    engine,
    "presto",
    "Engine to use for SQL evaluation. Valid values are 'presto' and 'spark'."
    "Default value is 'presto'.");

```
**Flag parsing**

```cpp
  std::cout << FLAGS_engine << std::endl;
  if (FLAGS_engine == "spark") {
    functions::sparksql::registerFunctions("");
  } else if (FLAGS_engine == "presto") {
    functions::prestosql::registerAllScalarFunctions();
    aggregate::prestosql::registerAllAggregateFunctions();
  } else {
    std::cout << "Invalid value for --engine: " << FLAGS_engine << ". ";
    std::cout << "Valid values are: 'presto' and 'spark'." << std::endl;
    exit(1);
  }


```

**Output after local testing**

<img width="534" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/dd1a975c-5924-49b3-8f03-a0899847721b">

If this sounds good, I will incorporate these changes into ExpressionRunnerTest.cpp and remove the print statement, spark_expression_runner_test executable, and new libraries this commit introduced.

codyschierbeck replied:
@kgpai 

My only concern is potential linking errors in the future. When I started trying to implement jsons for SparkSQL I ran into some linking errors. I think that may have just been because I was new to velox's build structure and C++ at the time. I shelved that work for a while, and when I came back to it after a little more experience I was able to avoid those errors. 

I am still not super familiar with the best practices in regards to avoiding naming collisions, etc, in this codebase, so maybe linking both function libraries together isn't an issue at all. Could use some insight from someone with more experience here.

kgpai reviewed (commented):
Thanks @codyschierbeck .

codyschierbeck commented:
@kgpai Hi, I implemented your requested changes. A new flag --engine is available for velox_expression_runner_test. It's default value is "presto". Can you please take a look?

kgpai commented on velox/expression/tests/ExpressionRunnerTest.cpp:
```diff
@@ -52,6 +53,12 @@ DEFINE_string(
+DEFINE_string(
+    engine,
+    "presto",
+    "Engine to use for expression evaluation. Currently supported values are "
```
Not sure if 'engine' is the right word, since all we are testing is the registry of functions. Maybe rename it to 'registry' ?

codyschierbeck replied:
Okay, I will that in the next batch of changes.

kgpai commented on velox/expression/tests/ExpressionRunnerTest.cpp:
```diff
@@ -181,9 +188,17 @@ int main(int argc, char** argv) {
+    functions::sparksql::registerFunctions("");
+  } else if (FLAGS_engine == "presto") {
+    functions::prestosql::registerAllScalarFunctions();
+    aggregate::prestosql::registerAllAggregateFunctions();
```
Do we need to register aggregate functions for expression fuzzer ?

codyschierbeck replied:
In Velox's current state, I believe the answer is no.

Taking a look at ExpressionFuzzerTest.cpp, PrestoSQL fuzzer doesn't seem to include aggregate tests. 

https://github.com/facebookincubator/velox/blob/5e56843165cd3f03ebf0ddd16196507e47ede4bb/velox/expression/tests/ExpressionFuzzerTest.cpp#L32-L59

So if we view this executables use as to only to debug Fuzzer failures, then no I would say it's best to exclude it. Additionally, velox_aggregates isn't even linked to velox_functions_prestosql or anything that velox_expression_fuzzer_test links to, so when you try to run sum for example, you get a core dump.

```
I0122 11:04:04.415263 1856286 ExpressionRunnerTest.cpp:130] File for complex_constant_path not found.
terminate called after throwing an instance of 'facebook::velox::VeloxUserError'
  what():  Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Scalar function doesn't exist: sum.
Retriable: False
Function: resolveScalarFunctionType
File: /home/sdp/pinterest/velox/velox/parse/TypeResolver.cpp
Line: 96
Stack trace:
Stack trace has been disabled. Use --velox_exception_user_stacktrace_enabled=true to enable it.

*** Aborted at 1705950244 (Unix time, try 'date -d @1705950244') ***
*** Signal 6 (SIGABRT) (0x3e8001c531e) received by PID 1856286 (pthread
```

The only reasons I can think for keeping it are:

1. If aggregates will eventually be included in the fuzzer
2. Or if Velox plans to add a way of creating custom on-disk vector representations. I have a personal local implementation of this as I've found it very helpful in my development, but I don't see a reason why Velox would ever support this by default

kgpai replied:
Thanks Cody,

1. We already have a fuzzer for aggregations, thus i think we can skip on registering aggregates.
2. Velox does have an on disk representation for Vectors, used purely for debugging (See VectorSaver) .

kgpai commented on velox/expression/tests/ExpressionRunnerTest.cpp:
```diff
@@ -181,9 +188,17 @@ int main(int argc, char** argv) {
+    functions::prestosql::registerAllScalarFunctions();
+    aggregate::prestosql::registerAllAggregateFunctions();
+  } else {
+    std::cout << "Invalid value for --engine: " << FLAGS_engine << ". ";
```
I think we have to use LOG(INFO) rather than cout here.

codyschierbeck replied:
Changing to LOG(INFO) means that by default when an incorrect value is passed executable exists without informing the user why unless they passed --v=1 and --logtostderr=1. 

I did modify this to model the other verifier for mode, an independant required argument:

```cpp
static bool validateMode(const char* flagName, const std::string& value) {
  static const std::unordered_set<std::string> kModes = {
      "common", "simplified", "verify", "query"};
  if (kModes.count(value) != 1) {
    std::cout << "Invalid value for --" << flagName << ": " << value << ". ";
    std::cout << "Valid values are: " << folly::join(", ", kModes) << "."
              << std::endl;
    return false;
  }

  return true;
}

static bool validateRegistry(const char* flagName, const std::string& value) {
  static const std::unordered_set<std::string> kRegistries = {
      "presto", "spark"};
  if (kRegistries.count(value) != 1) {
    std::cout << "Invalid value for --" << flagName << ": " << value << ". ";
    std::cout << "Valid values are: " << folly::join(", ", kRegistries) << "."
              << std::endl;
    return false;
  }
  if (value == "spark") {
    functions::sparksql::registerFunctions("");
  }
  else if (value == "presto") {
    functions::prestosql::registerAllScalarFunctions();
  }

  return true;
}

DEFINE_validator(mode, &validateMode);
DEFINE_validator(registry, &validateRegistry);
```

This also includes removing Presto aggregate registry.

kgpai replied:
You can use LOG(ERROR/WARN) when we absolutely need to output error or warning messages etc. We typically use std::cout only in example programs.

mbasmanova replied:
It seems reasonable to use std::cerr to print out the error message that explains why the program exited abnormally.

codyschierbeck replied:
The current standard in ExpressionRunner.cpp is to output flag misuse to std::cout. I'd be happy to modify the output for these as well if we prefer std::ceer.

https://github.com/facebookincubator/velox/blob/5fa64970fb9b6a9c7bf3103829c4132fc4f4942f/velox/expression/tests/ExpressionRunnerTest.cpp#L174-L177

https://github.com/facebookincubator/velox/blob/5fa64970fb9b6a9c7bf3103829c4132fc4f4942f/velox/expression/tests/ExpressionRunnerTest.cpp#L82-L93.

kgpai reviewed (commented):
We should also change all fuzzer jobs that use spark to explicitly invoke this flag after this change else they will use presto and be broken.

codyschierbeck commented:
> We should also change all fuzzer jobs that use spark to explicitly invoke this flag after this change else they will use presto and be broken.

I'm not clued in on the backend of CI/CD, but this isn't necessarily a fuzzer, right? It can be used to explain fuzzer failures without having to re-run the fuzzer, but from my understanding actual fuzzing is done by spark_expression_fuzzer_test and velox_expression_fuzzer_test, which already isolate to their respective function libraries. 

This was included because it wasn't possible to follow these docs: https://github.com/facebookincubator/velox/blob/main/velox/docs/develop/testing/fuzzer.rst#accurate-on-disk-reproduction for Spark functions. 

But yeah, I guess if there's something in the CI/CD pipeline that leverages the velox_expression_runner_test, those changes are necessary.

codyschierbeck commented:
@kgpai Outside of the linux-adapters test, everything here is passing. linux-adapters is failing because of the following:

```
288: unknown file: Failure
288: C++ exception with description "Exception: VeloxRuntimeError
288: Error Source: RUNTIME
288: Error Code: INVALID_STATE
288: Reason: Failed to get metadata for S3 object due to: 'Network connection'. Path:'s3://data3/test.txt', SDK Error Type:99, HTTP Status Code:-1, S3 Service:'Unknown', Message:'curlCode: 7, Couldn't connect to server', RequestID:''
288: Retriable: False
288: Function: initialize
288: File: ../../velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
288: Line: 93
288: Stack trace:
288: # 0  _ZN8facebook5velox7process10StackTraceC1Ei
288: # 1  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
288: # 2  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
288: # 3  _ZN8facebook5velox12_GLOBAL__N_110S3ReadFile10initializeEv
288: # 4  _ZN8facebook5velox11filesystems12S3FileSystem15openFileForReadESt17basic_string_viewIcSt11char_traitsIcEERKNS1_11FileOptionsE
288: # 5  _ZN8facebook5velox19FileHandleGeneratorclERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
288: # 6  _ZN8facebook5velox13CachedFactoryINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt10shared_ptrINS0_10FileHandleEENS0_19FileHandleGeneratorEE8generateERKS7_
288: # 7  _ZN8facebook5velox44S3FileSystemRegistrationTest_fileHandle_Test8TestBodyEv
288: # 8  _ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS_4TestEvEET0_PT_MS4_FS3_vEPKc
288: # 9  _ZN7testing4Test3RunEv
288: # 10 _ZN7testing8TestInfo3RunEv
288: # 11 _ZN7testing9TestSuite3RunEv
288: # 12 _ZN7testing8internal12UnitTestImpl11RunAllTestsEv
288: # 13 _ZN7testing8UnitTest3RunEv
288: # 14 main
288: # 15 __libc_start_main
288: # 16 _start
288: " thrown in the test body.
288: I0122 20:44:15.496309 93139 TempDirectoryPath.cpp:31] TempDirectoryPath:: removing all files from/tmp/velox_test_mGevHR
288: [  FAILED  ] S3FileSystemRegistrationTest.fileHandle (27675 ms)
...................

305/305 Test #300: velox_exec_test ........................................................   Passed  772.65 sec

99% tests passed, 1 tests failed out of 302

Total Test time (real) = 824.83 sec

The following tests did not run:
         53 - */MemoryAllocatorTest.allocContiguousVsize/* (Disabled)
        102 - */MemoryPoolTest.memoryLeakCheck/* (Disabled)
        103 - */MemoryPoolTest.growBeyondMaxCapacity/* (Disabled)

The following tests FAILED:
        288 - velox_s3registration_test (Failed)
Errors while running CTest

```
Outside of this, everything is passing. Can you take a look at the recent changes and let me know what you think?

mbasmanova commented:
@majetideepak Deepak, do you know if S3FileSystemRegistrationTest.fileHandle is flaky?

majetideepak commented:
> @majetideepak Deepak, do you know if S3FileSystemRegistrationTest.fileHandle is flaky?

`Failed to get metadata for S3 object due to: 'Network connection'. Path:'s3://data3/test.txt', SDK Error Type:99, HTTP Status Code:-1, S3 Service:'Unknown', Message:'curlCode: 7, Couldn't connect to server', RequestID:''`

`S3 Service:'Unknown'` could mean that the MinioServer started but was not ready to accept requests by the time this test ran. I have not seen this before, but probably now likely due to increased usage of the HW.
 Issue: https://github.com/facebookincubator/velox/issues/8488

kgpai commented:
@codyschierbeck We do run spark_expression_fuzzer_test and velox_expression_fuzzer_test for some nightly and scheduled runs . They are here : https://github.com/facebookincubator/velox/tree/main/.circleci and here https://github.com/facebookincubator/velox/tree/main/.github . We monitor these runs for bugs / errors , so having them fail is not ideal.

It makes sense to also change those places in this PR .

kgpai commented:
@codyschierbeck Apologies, I mistook these changes for expression_fuzzer and not expression_runner_* . expression_runner_* doesnt have any scheduled or nightly runs, so we do not need to change any yml files. 
It will still be good to edit the documentation here though : https://facebookincubator.github.io/velox/develop/testing/fuzzer.html#accurate-on-disk-reproduction .

codyschierbeck commented:
@kgpai @mbasmanova I believe this PR is ready.

To summarize, I linked velox_functions_spark to velox_expression_runner, added a flag to ExpressionRunnertest.cpp called registry that allows users to specify which function registry to use, and changed all output in that file from std::cout to std::cerr. 

Changed docs in fuzzer.rst to reflect this new flag. They state:

* ``--registry`` function registry to use for evaluating expression. One of "presto" (default) or "spark".

mbasmanova commented:
@kgpai Krishna, would you help review this PR?

codyschierbeck commented:
@kgpai Can you take a look?

codyschierbeck commented:
@kgpai Do you think this is ready for merge?

kgpai commented:
Thanks @codyschierbeck , Reviewing..

kevinwilfong commented on velox/expression/tests/ExpressionRunnerTest.cpp:
```diff
@@ -52,6 +53,12 @@ DEFINE_string(
+DEFINE_string(
+    registry,
+    "presto",
+    "Funciton registry to use for expression evaluation. Currently supported values are "
```
Sorry, one small nit, "Function" is misspelled here.

codyschierbeck replied:
Thank you! Great catch. Changing now

codyschierbeck replied:
Do you have access to the Linter failure? Any idea why it is failing?

kgpai commented:
@codyschierbeck Hoping to land this within  a day or so (need to do a minor internal fix).
